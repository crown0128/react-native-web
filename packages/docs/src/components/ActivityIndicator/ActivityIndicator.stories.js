import PropTypes from 'prop-types';

const ofProps = () => {};

ofProps.propTypes = {
  '...ViewPropTypes': PropTypes.any,
  animating: PropTypes.bool,
  color: PropTypes.string,
  hidesWhenStopped: PropTypes.bool,
  size: PropTypes.oneOfType([PropTypes.oneOf(['small', 'large']), PropTypes.number])
};

ofProps.defaultProps = {
  animating: true,
  color: '#1976D2',
  hidesWhenStopped: true,
  size: 'small'
};

export default {
  title: 'Components|ActivityIndicator',
  includeStories: []
};

export { ofProps };
export { default as animating } from './examples/Animating';
export { default as color } from './examples/Color';
export { default as hidesWhenStopped } from './examples/HidesWhenStopped';
export { default as size } from './examples/Size';
