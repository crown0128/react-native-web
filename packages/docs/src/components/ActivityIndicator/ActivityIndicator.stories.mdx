import { Meta, Story, Preview } from '@storybook/addon-docs/blocks';
import * as Stories from './examples';

<Meta title="Components|ActivityIndicator" />

# ActivityIndicator

Displays a customizable activity indicator.

## Props

| Name                      | Type      | Default |
| ------------------------- | --------- | ------- |
| ...ViewProps              | | |
| animating                 | ?boolean  | true |
| color                     | ?string   | '#1976D2' |
| hidesWhenStopped          | ?boolean  | true |
| size                      | 'small' \| 'large' \| number | 'small' |

### animating

Controls whether to show the indicator or hide it.

<Preview withSource='none'>
  <Story name="animating">
    <Stories.animating />
  </Story>
</Preview>

### color

Customize the foreground color of the indicator.

<Preview withSource='none'>
  <Story name="color">
    <Stories.color />
  </Story>
</Preview>

### hidesWhenStopped

Control whether to show the indicator when it is not animating.

<Preview withSource='none'>
  <Story name="hidesWhenStopped">
    <Stories.hidesWhenStopped />
  </Story>
</Preview>

### size

Customize the size of the indicator. `small` has a height of `20px`, large has a
height of `36px`. Scale transforms can also be used.

<Preview withSource='none'>
  <Story name="size">
    <Stories.size />
  </Story>
</Preview>
