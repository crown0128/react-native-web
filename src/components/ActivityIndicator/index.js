import Animated from '../../apis/Animated';
import applyNativeMethods from '../../modules/applyNativeMethods';
import Easing from 'animated/lib/Easing';
import StyleSheet from '../../apis/StyleSheet';
import View from '../View';
import React, { Component, PropTypes } from 'react';

const DEFAULT_COLOR = '#1976D2';
const rotationInterpolation = { inputRange: [ 0, 1 ], outputRange: [ '0deg', '360deg' ] };

class ActivityIndicator extends Component {
  static propTypes = {
    animating: PropTypes.bool,
    color: PropTypes.string,
    hidesWhenStopped: PropTypes.bool,
    size: PropTypes.oneOf([ 'small', 'large' ]),
    style: View.propTypes.style
  };

  static defaultProps = {
    animating: true,
    color: DEFAULT_COLOR,
    hidesWhenStopped: true,
    size: 'small',
    style: {}
  };

  constructor(props) {
    super(props);
    this.state = {
      animation: new Animated.Value(0)
    };
  }

  componentDidMount() {
    this._manageAnimation();
  }

  componentDidUpdate() {
    this._manageAnimation();
  }

  render() {
    const {
      animating,
      color,
      hidesWhenStopped,
      size,
      style,
      ...other
    } = this.props;

    const { animation } = this.state;

    const svg = (
      <svg height='100%' viewBox='0 0 32 32' width='100%'>
        <circle
          cx='16'
          cy='16'
          fill='none'
          r='14'
          strokeWidth='4'
          style={{
            stroke: color,
            opacity: 0.2
          }}
        />
        <circle
          cx='16'
          cy='16'
          fill='none'
          r='14'
          strokeWidth='4'
          style={{
            stroke: color,
            strokeDasharray: 80,
            strokeDashoffset: 60
          }}
        />
      </svg>
    );

    return (
      <View {...other} style={[ styles.container, style ]}>
        <Animated.View
          children={svg}
          style={[
            indicatorStyles[size],
            hidesWhenStopped && !animating && styles.hidesWhenStopped,
            {
              transform: [
                { rotate: animation.interpolate(rotationInterpolation) }
              ]
            }
          ]}
        />
      </View>
    );
  }

  _manageAnimation() {
    const { animation } = this.state;

    const cycleAnimation = () => {
      animation.setValue(0);
      Animated.sequence([
        Animated.timing(animation, {
          duration: 750,
          easing: Easing.inOut(Easing.linear),
          toValue: 1
        })
      ]).start((event) => {
        if (event.finished) {
          cycleAnimation();
        }
      });
    };

    if (this.props.animating) {
      cycleAnimation();
    } else {
      animation.stopAnimation();
    }
  }
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center'
  },
  hidesWhenStopped: {
    visibility: 'hidden'
  }
});

const indicatorStyles = StyleSheet.create({
  small: {
    width: 20,
    height: 20
  },
  large: {
    width: 36,
    height: 36
  }
});

module.exports = applyNativeMethods(ActivityIndicator);
