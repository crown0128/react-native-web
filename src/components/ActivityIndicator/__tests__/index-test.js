/* eslint-env jasmine, jest */

import ActivityIndicator from '..';
import React from 'react';
import renderer from 'react-test-renderer';

describe('components/ActivityIndicator', () => {
  test('default render', () => {
    const component = renderer.create(<ActivityIndicator />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('other render', () => {
    const component = renderer.create(
      <ActivityIndicator animating={false} hidesWhenStopped={false} size="large" />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
