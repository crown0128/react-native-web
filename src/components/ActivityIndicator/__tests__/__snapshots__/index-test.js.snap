exports[`components/ActivityIndicator default render 1`] = `
<div
  aria-valuemax="1"
  aria-valuemin="0"
  className="rn-alignItems-1awozwy rn-backgroundColor-wib322 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-font-1lw9tu2 rn-justifyContent-1777fci rn-listStyle-1ebb2ja rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-textAlign-1ttztb7 rn-textDecoration-bauka4"
  role="progressbar"
  style={Object {}}>
  <div
    className="rn-alignItems-1oszu61 rn-animationDuration-17bb2tj rn-animationIterationCount-1muvv40 rn-animationName-dozj4v rn-animationTimingFunction-1ldzwu0 rn-backgroundColor-wib322 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-font-1lw9tu2 rn-height-z80fyv rn-listStyle-1ebb2ja rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-textAlign-1ttztb7 rn-textDecoration-bauka4 rn-width-19wmn03"
    style={Object {}}>
    <svg
      height="100%"
      viewBox="0 0 32 32"
      width="100%">
      <circle
        cx="16"
        cy="16"
        fill="none"
        r="14"
        strokeWidth="4"
        style={
          Object {
            "opacity": 0.2,
            "stroke": "#1976D2",
          }
        } />
      <circle
        cx="16"
        cy="16"
        fill="none"
        r="14"
        strokeWidth="4"
        style={
          Object {
            "stroke": "#1976D2",
            "strokeDasharray": 80,
            "strokeDashoffset": 60,
          }
        } />
    </svg>
  </div>
</div>
`;

exports[`components/ActivityIndicator other render 1`] = `
<div
  aria-valuemax="1"
  aria-valuemin="0"
  className="rn-alignItems-1awozwy rn-backgroundColor-wib322 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-font-1lw9tu2 rn-justifyContent-1777fci rn-listStyle-1ebb2ja rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-textAlign-1ttztb7 rn-textDecoration-bauka4"
  role="progressbar"
  style={Object {}}>
  <div
    className="rn-alignItems-1oszu61 rn-animationDuration-17bb2tj rn-animationIterationCount-1muvv40 rn-animationName-dozj4v rn-animationPlayState-1abnn5w rn-animationTimingFunction-1ldzwu0 rn-backgroundColor-wib322 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-font-1lw9tu2 rn-height-1r8g8re rn-listStyle-1ebb2ja rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-textAlign-1ttztb7 rn-textDecoration-bauka4 rn-width-1acpoxo"
    style={Object {}}>
    <svg
      height="100%"
      viewBox="0 0 32 32"
      width="100%">
      <circle
        cx="16"
        cy="16"
        fill="none"
        r="14"
        strokeWidth="4"
        style={
          Object {
            "opacity": 0.2,
            "stroke": "#1976D2",
          }
        } />
      <circle
        cx="16"
        cy="16"
        fill="none"
        r="14"
        strokeWidth="4"
        style={
          Object {
            "stroke": "#1976D2",
            "strokeDasharray": 80,
            "strokeDashoffset": 60,
          }
        } />
    </svg>
  </div>
</div>
`;
